internal/controller/apiextensions/composite/patches.go

```diff
@@ -0,0 +1,308 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package composite
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+)
+
+const (
+	errPatchSetType             = "a patch in a PatchSet cannot be of type PatchSet"
+	errCombineRequiresVariables = "combine patch types require at least one variable"
+
+	errFmtUndefinedPatchSet           = "cannot find PatchSet by name %s"
+	errFmtInvalidPatchType            = "patch type %s is unsupported"
+	errFmtCombineStrategyNotSupported = "combine strategy %s is not supported"
+	errFmtCombineConfigMissing        = "given combine strategy %s requires configuration"
+	errFmtCombineStrategyFailed       = "%s strategy could not combine"
+	errFmtExpandingArrayFieldPaths    = "cannot expand ToFieldPath %s"
+)
+
+// Apply executes a patching operation between the from and to resources.
+// Applies all patch types unless an 'only' filter is supplied.
+func Apply(p v1.Patch, cp resource.Composite, cd resource.Composed, only ...v1.PatchType) error {
+	if filterPatch(p, only...) {
+		return nil
+	}
+
+	switch p.Type {
+	case v1.PatchTypeFromCompositeFieldPath:
+		return ApplyFromFieldPathPatch(p, cp, cd)
+	case v1.PatchTypeToCompositeFieldPath:
+		return ApplyFromFieldPathPatch(p, cd, cp)
+	case v1.PatchTypeCombineFromComposite:
+		return ApplyCombineFromVariablesPatch(p, cp, cd)
+	case v1.PatchTypeCombineToComposite:
+		return ApplyCombineFromVariablesPatch(p, cd, cp)
+	case v1.PatchTypePatchSet:
+		// Already resolved - nothing to do.
+	}
+	return errors.Errorf(errFmtInvalidPatchType, p.Type)
+}
+
+// filterPatch returns true if patch should be filtered (not applied)
+func filterPatch(p v1.Patch, only ...v1.PatchType) bool {
+	// filter does not apply if not set
+	if len(only) == 0 {
+		return false
+	}
+
+	for _, patchType := range only {
+		if patchType == p.Type {
+			return false
+		}
+	}
+	return true
+}
+
+// ResolveTransforms applies a list of transforms to a patch value.
+func ResolveTransforms(c v1.Patch, input any) (any, error) {
+	var err error
+	for i, t := range c.Transforms {
+		if input, err = Resolve(t, input); err != nil {
+			// TODO(negz): Including the type might help find the offending transform faster.
```

Comment:
@ezgidemirel fyi, I believe @negz preferred to leave this as a to-do to keep the PR focused, but this would be a DX improvement to identify composition errors.